add Gadget2Particle

elehcim opened the pull request:
- Many routines translated from pynbody to read gadget2 files (format 1 and 2).
- Drastically reduced code repetition.

TODO:
- [ ] A config file needed for collecting gadget user defined fields and relative units.
- [x] `StructArray` data type (e.g. `Float32` of `Float64`) can't be instantiated at runtime. I think it should be possible, but still I need to find a way.
- [x]  Rebase with latest modifications upstream
- [ ] Implement write operations with new block based mechanics
- [x] Implement Mass block header dependant read/write operation

I pushed because I saw you're very nicely going forward. I did this work in the past two weeks but newly arrived commitments hindered me to finishing this up. By having it here, you can have a look and probably improve or pick up something for the codebase. I'll try to keep this going in any case.

The main architecture change is the introduction of a `Gadget2Block` type which detects and reads 1D and 3D blocks and knows about the datatype (`Float32`, `Float64`, `Int`) of the data block itself. This uses some routines from pynbody and reduces a lot of code repetition.
Then a newly created `Gadget2Particle` struct is added containing fields commonly used in Gadget2 files. This type is then used to load a `StructArray` and fill it up with a common file-reading machinery on blocks.

elehcim committed:
add Gadget2Particle as it is. WIP - request for comments and help.

- Many routines translated from pynbody to read gadget2 files (format 1 and 2).
- Drastically reduced code repetition.

TODO:
- A config file needed for collect gadget user defined fields and relative units.
- StructArray data type (e.g. Float32 of Float64) can't be instantiated at runtime. I think it should be possible, but still I need to find a way.
- Rebase with latest modifications upstream

islent commented:
Thanks! I will review the code next week (soon after I have updated documentations and release packages of JuliaAstroSim). Before that, AstroIO will release a minor patch to solve some compat problems.

> `StructArray` data type (e.g. `Float32` of `Float64`) can't be instantiated at runtime. I think it should be possible, but still I need to find a way.

My clumsy solution is to construct a small `StructArray` and then empty it:
https://github.com/JuliaAstroSim/AstroIO.jl/blob/901d337ea63448d12142e42df7b8ad2c840e91f2/src/Gadget.jl#L285-L289

> gadget user defined fields and relative units.

I decide to force user to determine what units they are using while reading snapshots, for example,
```julia
header, data = read_gadget2("snapshot.gadget2", uAstro)
```
where argument `uAstro` is required thereafter. And the units in snapshot file can be change by
```julia
header, data = read_gadget2("snapshot.gadget2", uAstro, uGadget2) #! There will be unit conversions

write_gadget2("output.Gadget2", header, data, uGadget2)
```
where `uGadget2` is the default units for I/O.

elehcim committed:
Fix data initialization and datatypes

More elegant way, not having to append!
(which was causing a 'type does not have a definite number of fields' error).
Also probably more efficient because memory is contiguous from the beginning.

islent commented:
> initializing a full-length StructArray and then write the proper Collection value in the correct positions

Wow! Great idea!

As for the conf file, [ConfParser.jl](https://github.com/JuliaIO/ConfParser.jl) should be helpful. I had thought about it once before, however it's hard to make the whole interface user-friendly. `pynbody` might be a good reference.

My idea is that, users can choose to
1. write a conf file (AstroIO will construct a handler (struct, Dict or Array) that handles all information. `POS`, `:Position`, `"kpc"` are packed in a struct)
2. construct the handler manually

elehcim commented:
Ok, thanks. I'll have a look at ConfParser.jl

islent commented:
Are you pushing to https://github.com/elehcim/AstroIO.jl? Where can I contribute some help?

elehcim commented:
Now I have a problem regarding unit conversion which seems to change the datatype.

For example:

```julia
source_units = AstroIO.get_units(:Vel, uGadget2);
target_units = AstroIO.get_units(:Vel, uAstro);
v = PVector(3.f0, 3.f0, 4.f0, source_units);
println(typeof(uconvert(target_units, v.x)))
     Quantity{Float64, 𝐋 𝐓^-1, Unitful.FreeUnits{(Gyr^-1, kpc), 𝐋 𝐓^-1, nothing}}
```
but I'd like it to stay a Float32.

I don't know now the solution. Do you think a type stable uconvert for PVector would be a solution? Any idea?

elehcim commented:
> Are you pushing to https://github.com/elehcim/AstroIO.jl? Where can I contribute some help?

Yes, I'm pushing there.
You're very welcome to contrbute, don't know exactly what is the best way :D. This PR allows maintainers to edit.

elehcim committed:
Introduce units and fileunits

elehcim committed:
Fix bug, now GadgetFormat1 tests succeed

elehcim committed:
Introduce unitless Gadget2Particle with Nothing as units

elehcim committed:
add size check on format2

elehcim committed:
use uEntropy

elehcim committed:
Skip unknown blocks for now

elehcim committed:
fix tests

elehcim committed:
Add combinatorics dependency

elehcim committed:
add wrapper to read_gadget. Comments and remove debug lines

elehcim committed:
Change peek to read

because there is no method for
`peek(::Stream{DataFormat{:Gadget2}, args...)`

elehcim committed:
manage MASS block header data

elehcim committed:
little fixes and add tests

elehcim committed:
remove unused functions

elehcim committed:
Use Gadget2Stream

elehcim committed:
do-block syntax

elehcim committed:
import only the required function from Combinatorics

elehcim committed:
Fix and add tests for reading mass from header

elehcim commented:
How can I infer the numeric type from a `PVector`?
Example: 
```julia
PVector{Quantity{Float32, 𝐋, Unitful.FreeUnits{(kpc,), 𝐋, nothing}}}  --> Float32
```

to be used like:
```julia
julia> infer_numeric_type(eltype(data.Pos))
Float32
```

elehcim commented:
> How can I infer the numeric type from a `PVector`? Example:
> 
> ```julia
> PVector{Quantity{Float32, 𝐋, Unitful.FreeUnits{(kpc,), 𝐋, nothing}}}  --> Float32
> ```
> 
> to be used like:
> 
> ```julia
> julia> infer_numeric_type(eltype(data.Pos))
> Float32
> ```

For now I'll do:
```julia
eltype(d.Pos).parameters[1].parameters[1]
```

islent commented:
> `eltype(d.Pos).parameters[1].parameters[1]`

I get a better solution (in the latest master branch of `PhysicalParticles`):
https://github.com/JuliaAstroSim/PhysicalParticles.jl/blob/4114ddc0e4ac959fa8c6cfe9c7275f5662ed6983/src/PVector.jl#L198-L205

Have fun with it :D
```julia
julia> using PhysicalParticles

julia> a = PVector(1.0f0, 2.0f0, 3.0f0)
PVector{Float32}(1.0f0, 2.0f0, 3.0f0)

julia> numeric_type(a)
Float32

julia> numeric_type([a])
Float32
```

islent commented:
Merged through https://github.com/JuliaAstroSim/AstroIO.jl/pull/12. (Sorry that I was not familiar with editing PRs before merging. I should have merged this PR first and then commit to master branch)

> A config file needed for collecting gadget user defined fields and relative units.

I have implemented a config file loading procedure in `ConfParser.jl`:
1. load a config file with field `[io]`
2. define `IOConfig`
3. load snapshot using `IOConfig`

There is an example of loading `JLD2` file:
https://github.com/JuliaAstroSim/AstroIO.jl/blob/1bb9949c95c85aad20ba627c51d6ab3ede311553/test/runtests.jl#L142-L145

How to define blocks from `iofields` (which is a `Vector` of 4-Char strings, for example, `["POS ", "VEL "...]`)?
https://github.com/JuliaAstroSim/AstroIO.jl/blob/1bb9949c95c85aad20ba627c51d6ab3ede311553/src/ConfParser.jl#L3-L7

islent committed:
Merge branch 'master' into Gadget2Particle

islent commented:
I'm also planing to remove SPH fields from `PhysicalParticles.Star` and migrate `Gadget2Particle` to `PhysicalParticles`. Will these cause any errors?

elehcim commented:
Hi!

> Merged through #12. (Sorry that I was not familiar with editing PRs before merging. I should have merged this PR first and then commit to master branch)

No problem, thanks for merging. I think I'll close this one and open a new one dedicated to the write operations. And also pushing forward the config file.
I'm thinking to use the [StructTypes.jl model](https://juliadata.github.io/StructTypes.jl/stable/#StructTypes.constructfrom).


> > A config file needed for collecting gadget user defined fields and relative units.
> 
> I have implemented a config file loading procedure in `ConfParser.jl`:
> 
> 1. load a config file with field `[io]`
> 2. define `IOConfig`
> 3. load snapshot using `IOConfig`
> 

Thanks, I'll look into it.

> There is an example of loading `JLD2` file:
> 
> https://github.com/JuliaAstroSim/AstroIO.jl/blob/1bb9949c95c85aad20ba627c51d6ab3ede311553/test/runtests.jl#L142-L145
> 
> How to define blocks from `iofields` (which is a `Vector` of 4-Char strings, for example, `["POS ", "VEL "...]`)?
> 
> https://github.com/JuliaAstroSim/AstroIO.jl/blob/1bb9949c95c85aad20ba627c51d6ab3ede311553/src/ConfParser.jl#L3-L7

elehcim commented:
> I'm also planing to remove SPH fields from `PhysicalParticles.Star` and migrate `Gadget2Particle` to `PhysicalParticles`. Will these cause any error?

I don't think so. But I'm thinking that in `PhysicalParticles.jl` there should be a struct which is your interface to your simulation code.
I see `Gadget2Particle` as a struct for post processing simulation data.

It may contain what is there in the gadget2 file (guided by the config file).

If we convene to use `StructTypes.jl`, it actually may contain many custom fields generated on the fly (if I undertand well the [contructfrom](https://juliadata.github.io/StructTypes.jl/stable/#StructTypes.constructfrom) method).
Also StructTypes.jl has a nice architecture to select serializable fields.
 
Let me know what do you think.

elehcim commented:
Ah, after commit b655ee3, I think I can continue here.

islent commented:
> But I'm thinking that in PhysicalParticles.jl there should be a struct which is your interface to your simulation code.

Yep. I'm using `Star` for gravity-only simulations. So SPH info is redundant for this.

There are only two restrictions to user-defined particle types for a N-body force solver:
1. Be a subtype of `AbstractParticle`
2. Contain at least the same data fields in `Star`

So It's ok to leave `Gadget2Particle` to `AstroIO`. And the situation is same with `StructTypes.jl` (just to emphasize it, use immutable struct for higher performance)
